### PR TITLE
Add ThroughField to import

### PR DESCRIPTION
**Pull request by szok**
TODO
- [x] unittests

### CODE REVIEW

File: src/ralph/data_importer/fields.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        super().__init__(attribute, column_name, widget, readonly)
+
+    def save(self, obj, data):
+        if not self.readonly:
+            value = data.get(self.column_name)
+            values = self.widget.clean(value)
+            if values:
+                self.through_class.objects.filter(
+                    **{self.through_field_obj: obj}
+                ).delete()
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py:68]: check how to do this without deleting (more updating way)

File: src/ralph/data_importer/fields.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        self.through_class = through_class
+        self.through_field_obj = through_field_obj
+        self.through_field = through_field
+        super().__init__(attribute, column_name, widget, readonly)
+
+    def save(self, obj, data):
+        if not self.readonly:
+            value = data.get(self.column_name)
+            values = self.widget.clean(value)
+            if values:
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: remove this if - if it is empty, all values should be deleted

File: src/ralph/data_importer/widgets.py [hunk 3]
```diff
@@ -79,6 +80,35 @@ def render(self, value):
         return self.separator.join([obj.username for obj in value.all()])
 
 
+class LicenceUserWidget(widgets.ManyToManyWidget):
+
+    def clean(self, value):
+        if not value:
+            return get_user_model().objects.none()
+        usernames = value.split(self.separator)
+        return get_user_model().objects.filter(username__in=usernames)
```
Comment (mkurek) [inline src/ralph/data_importer/widgets.py:89]: shouldn't we warn user that some of the users (values) were not found?

File: src/ralph/data_importer/fields.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        self, through_class, through_field_obj, through_field,
+        attribute=None, column_name=None, widget=None,
+        readonly=False
+    ):
+        self.through_class = through_class
+        self.through_field_obj = through_field_obj
+        self.through_field = through_field
+        super().__init__(attribute, column_name, widget, readonly)
+
+    def save(self, obj, data):
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: add transaction

File: src/ralph/data_importer/fields.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            value = data.get(self.column_name)
+            old_objs = set(self.through_class.objects.all())
+            for i in self.widget.clean(value):
+                through_obj, _ = self.through_class.objects.get_or_create(
+                    **{self.through_field_obj: obj, self.through_field: i}
+                )
+                if through_obj in old_objs:
+                    old_objs.remove(through_obj)
+
+            for i in old_objs:
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: can't you delete then in bulk?

File: src/ralph/data_importer/tests/test_fields.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            through_field_obj='licence',
+            through_field='user',
+            attribute='users',
+            column_name='users',
+            widget=LicenceUserWidget(model=get_user_model())
+        )
+
+        self.assertEqual(self.licence.users.all().count(), 1)
+
+        field.save(
```
Comment (mkurek) [inline src/ralph/data_importer/tests/test_fields.py]: could you make assertion on number of queries here? https://docs.djangoproject.com/en/dev/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries

File: src/ralph/data_importer/fields.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
+from import_export import fields
+
+
+class ThroughField(fields.Field):
+
+    def __init__(
+        self, through_class, through_field_obj, through_field,
+        attribute=None, column_name=None, widget=None,
+        readonly=False
+    ):
```
Comment (ghost) [inline src/ralph/data_importer/fields.py:10]: I'd add here docstring with detailed desc. of parameters.

File: src/ralph/data_importer/widgets.py (hunk 3 above)
Comment (ghost) [inline src/ralph/data_importer/widgets.py:89]: :+1:

File: src/ralph/data_importer/widgets.py [hunk 8]
```diff
@@ -79,6 +80,35 @@ def render(self, value):
...
+        )
+
+
+class BaseObjectLicenceWidget(widgets.ManyToManyWidget):
+
+    def clean(self, value):
+        if not value:
+            return BaseObject.objects.none()
+        return BaseObject.objects.filter(
+            pk__in=value.split(self.separator)
```
Comment (ghost) [inline src/ralph/data_importer/widgets.py:103]: perhaps warn. here also wanted?

File: src/ralph/data_importer/fields.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Field for through django model
+
+        Args:
+            through_class: Django through class
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: Django through model for M2M relation

File: src/ralph/data_importer/fields.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            through_field_obj: related field for model
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: is this really field?

File: src/ralph/data_importer/fields.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            through_field: related field to selected model
+            attribute: string of either an instance attribute or callable
+                off the object
+            column_name: let you provide how this field is named
+                in datasource.
+            widget: efines widget that will be used to represent field
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: defines

File: src/ralph/data_importer/fields.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+                data in export
+            readonly: boolean value defines that if this field will
+                be assignedto object during import
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: assigned to

File: src/ralph/data_importer/tests/test_fields.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+from ralph.data_importer.resources import LicenceResource
+from ralph.data_importer.widgets import LicenceUserWidget
+from ralph.licences.models import Licence, LicenceUser
+from ralph.licences.tests.factories import LicenceFactory
+
+
+class DataImporterFieldsTestCase(TestCase):
+    def setUp(self):  # noqa
+        self.licence = LicenceFactory()
+        self.users = [UserFactory() for i in range(2)]
```
Comment (mkurek) [inline src/ralph/data_importer/tests/test_fields.py]: create more users (ex. 4) here - number of sql queries shouldnt change

File: src/ralph/data_importer/fields.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        self.through_class = through_class
+        self.through_field_obj = through_field_obj
+        self.through_field = through_field
+        super().__init__(attribute, column_name, widget, readonly)
+
+    def save(self, obj, data):
+        if not self.readonly:
+            value = data.get(self.column_name)
+            current = set(self.widget.clean(value))
```
Comment (mkurek) [inline src/ralph/data_importer/fields.py]: add some comment where there are objects and where just ids

File: src/ralph/data_importer/tests/test_fields.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+from ralph.data_importer.widgets import LicenceUserWidget
+from ralph.licences.models import Licence, LicenceUser
+from ralph.licences.tests.factories import LicenceFactory
+
+
+class DataImporterFieldsTestCase(TestCase):
+    def setUp(self):  # noqa
+        self.licence = LicenceFactory()
+        self.users = [UserFactory() for i in range(4)]
+        LicenceUser.objects.create(
```
Comment (mkurek) [inline src/ralph/data_importer/tests/test_fields.py:16]: please add one additional `LicenceUser` object

File: src/ralph/data_importer/tests/test_fields.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            through_from_field_name='licence',
+            through_to_field_name='user',
+            attribute='users',
+            column_name='users',
+            widget=LicenceUserWidget(model=get_user_model())
+        )
+
+        self.assertEqual(self.licence.users.all().count(), 1)
+
+        with self.assertNumQueries(4):
```
Comment (mkurek) [inline src/ralph/data_importer/tests/test_fields.py:37]: imo there should be max 4 queries:
- `self.widget.clean(value)`
- `self.through_model.objects.all()`
- `self.through_model.objects.bulk_create(to_add_list)`
- `self.through_model.objects.filter(...).delete()`

but in your case there is any remove, so there should be 3

Could you remove one `LicenceUser` in this test too?

**Comment by mkurek** [comment]
:ok:

**Comment by ar4s** [comment]
:ok:

**Pull request merged**